[Infrastructure UI] Wrap only the snapshot page with waffle providers

jennypavlova opened the pull request:
Closes #144387
This PR adds the waffle URL state only to the inventory page. So the waffle URL parameters shouldn't be visible on the other infra pages where those parameters are not used. One downside is when the user navigates to the host view and back to inventory the waffle options are not persisted.

Note: All waffle state parameters (filters, time, options) are needed in the locale storage but only for the time I added the `UrlStateContainer` because the filters and option are already set in a context on the upper level ( in `inventoryPrefill` to `useAlertPrefillContext`) 
❓  I don't know why the time is not set in this context (probably the time it's not needed there as it's the alerting context). That's why I added  [use_waffle_time_url_state.tsx](https://github.com/elastic/kibana/blob/45eebceae0b8634d448723c5e2ec4127dc32d351/x-pack/plugins/infra/public/pages/metrics/inventory_view/hooks/use_waffle_time_url_state.tsx) to make it work when the time is changed.

Notes for testing:
The waffle URL state should only show on the inventory page so on the host/metrics explorer page for example any of the waffle state parameters shouldn't be visible.
⚠️ The saved view should not be selected/saved as the page will load the saved data from the saved view
<img width="2304" alt="image" src="https://user-images.githubusercontent.com/14139027/200384386-54dad146-7d74-4f8b-9d3b-4b844105636c.png">

elasticmachine commented:
Pinging @elastic/infra-monitoring-ui (Team:Infra Monitoring UI)

miltonhultgren commented:
> One downside is when the user navigates to the host view and back to inventory the waffle options are not persisted.

I think the metric explorer puts it's state into localStorage for this reason so it can re-hydrate the last view, but I think there are better Kibana primitives for this as we move into using Unified Search in Metrics as well?

jennypavlova commented:
I tried using the locale storage (for the waffle time for now) similar to metrics explorer and the time value is set correctly there but when loading the data view it will override the current URL state. It is also something that happens on `main` currently for the time value - when switching to another tab and back to the inventory the URL state is also ot persisted. 

In metrics explorer we don't load the `default data view` by default - there the default is `no view selected` and in inventory, we load the `default data view` by default. So this could be something that is causing the override. 

I will think about a possible solution for the `currentTime` value inside the waffle time or I will revert the local storage changes and try a different approach.

neptunian commented:
I think you are using "Data Views" to refer to "Saved Views".  "Data Views" is a [different thing](https://www.elastic.co/guide/en/kibana/master/data-views.html).

I would not expect my UI changes to persist from switching from app to app (except unified search component).  I've checked some other apps including Logs and it does not.  So I don't think I'd worry about this.  Considering I can create a Saved View and then set that Saved View as my Default View, it seems unnecessary.  Can leave Metrics Explorer local storage state as is, though.

jennypavlova commented:
> I think you are using "Data Views" to refer to "Saved Views". "Data Views" is a [different thing](https://www.elastic.co/guide/en/kibana/master/data-views.html).

Ah sorry about that, I think I saw data view somewhere or maybe I got confused - I updated the description so it's not confusing.

> I would not expect my UI changes to persist from switching from app to app (except unified search component). I've checked some other apps including Logs and it does not. So I don't think I'd worry about this. Considering I can create a Saved View and then set that Saved View as my Default View, it seems unnecessary. 

I am trying to still keep the old behavior and it persisted only between infrastructure pages before when I tested it (because the context was used on the Infrastructure routing level) - you are right, a saved view can be created if needed (also the URL can be used/bookmarked to restore the filters) 

> Can leave Metrics Explorer local storage state as is, though.

I am not sure that I fully understand what you mean - I only extracted the function used to set the state to the locale storage from metrics explorer so I can reuse it for the waffle state, I don't think I changed anything else there or I am missing something 🤔

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/inventory_view/hooks/use_waffle_time_url_state.tsx:
```diff
@@ -0,0 +1,52 @@
+  );
+};
+
+export const mapToUrlState = (value: any): WaffleTimeState | undefined => {
```
`value` could have the following type:
```
{
  currentTime: number(?)
  isAutoReload: boolean
}
```

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/inventory_view/hooks/use_waffle_time_url_state.tsx:
```diff
@@ -0,0 +1,52 @@
+};
+
+export const mapToUrlState = (value: any): WaffleTimeState | undefined => {
+  if (value?.currentTime) {
+    const { currentTime, isAutoReloading } = value;
+    return { currentTime, isAutoReloading };
+  }
+  return undefined;
```
nit (maybe improves readability?): 
```suggestion
  if (!value?.currentTime) {
    return undefined
  }
   const { currentTime, isAutoReloading } = value;
   return { currentTime, isAutoReloading };
```

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/inventory_view/hooks/use_waffle_time.test.tsx:
```diff
@@ -0,0 +1,86 @@
+/*
```
❤️ !

crespocarlos commented on x-pack/plugins/infra/public/lib/settings_locale_storage.ts:
```diff
@@ -0,0 +1,31 @@
+
+import { Dispatch, SetStateAction, useEffect, useState } from 'react';
+
+function parseJsonOrDefault<Obj>(value: string | null, defaultValue: Obj): Obj {
```
nit (more common to see generic type names as T or TSomething):
```suggestion
function parseJsonOrDefault<TObj>(value: string | null, defaultValue: TObj): TObj {
```

crespocarlos reviewed (commented):
I ran a couple of tests and everything works as expected. Just left a few comments. Great job moving the URL state variables to the right place.

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/index.tsx:
```diff
@@ -42,6 +42,7 @@ import { InfraMLCapabilitiesProvider } from '../../containers/ml/infra_ml_capabi
 import { AnomalyDetectionFlyout } from './inventory_view/components/ml/anomaly_detection/anomaly_detection_flyout';
 import { HeaderActionMenuContext } from '../../utils/header_action_menu_provider';
 import { CreateDerivedIndexPattern } from '../../containers/metrics_source';
+import { WithWaffleTimeUrlState } from './inventory_view/hooks/use_waffle_time_url_state'; // WIP Time only
```
Do we need to keep this `WIP` comment?

jennypavlova commented:
While testing the changes with @neptunian we found some unusual behavior around saved views. Also on `main` when loading inventory the default view is preselected sometimes (even without changing the filters). So we decided to put this PR on hold for now. It's not an important fix for now and I am afraid that it can affect the saved views and break the expected behavior there. 
@crespocarlos Thank you for the review anyway! We found the issue after that and I am sorry that I couldn't ping you beforehand.
